### DISCUSSION TITLE
Surprising behavior when narrowing unions containing `Unknown`

### DISCUSSION THREAD

**Discussion by nelhage** [microsoft/pyright#5592] [Q&A] [answered]
I'm surprised by the behavior when running pyright on this snippet:

```python
def some_prefix(x) -> bool: ...

def f(x):
    if some_prefix(x):
        x = dict(key="hi", other="there")

    reveal_type(x)       # Type of "x" is "dict[str, str] | Unknown"

    if isinstance(x, dict):
        # THIS IS THE SURPRISING TYPE:
        reveal_type(x)   # Type of "x" is "dict[str, str]"
```

Inside the `if`, we see that the `Unknown` has been entirely narrowed away; I would have expected us to instead narrow to `dict[Unknown, Unknown]` or possibly `dict[Unknown, Unknown] | dict[str, str]`, since there's no reason `x` couldn't originally have been an arbitrary dict type. 

This is the behavior I see if I instead explicitly type `x: Any` in the signature: the `if` narrows to `dict[Unknown, Unknown]`. _A priori_ I expect `Unknown` and `Any` to behave virtually identically, at least in situations like this.

Is this a bug? A surprising consequence of intended behavior? A design choice somewhere to make `Unknown` and `Any` different in this way?

**Comment by erictraut** [comment, marked as answer]
Yes, this is intended behavior. The `isinstance` type guard needs to assumptions to trade off type correctness and practicality. The current behavior is the result of a few years of fine tuning based on user input.

You'll see the same behavior for narrowing `dict[str, str] | Any`.

```python
def f(x: dict[str, str] | Any):
    if isinstance(x, dict):
        reveal_type(x)  # dict[str, str]
```

Consider the following:

```python
class Foo: ...

def func(x: dict[str, str] | Foo):
    if isinstance(x, dict):
        reveal_type(x)
```

What would you expect the revealed type to be in this case? If pyright favored strictness over practicality, the answer would be `dict[str, str] | dict[Unknown, Unknown]` because the type of `x` could be derived from `Foo` and `dict`. Practically speaking, this isn't what pyright users expect.

For the same reason, pyright narrows `dict[str, str] | Unknown` (or `dict[str, str] | Any`) to `dict[str, str]` in this case.

**Comment by nelhage** [comment]
> or `dict[str, str] | Any`

Ah! I had observed that I didn't see the same narrowing behavior with `x: Any` in:

```python
def f(x):
    if some_prefix(x: Any):
        x = dict(key="hi", other="there")

    reveal_type(x)

    if isinstance(x, dict):
        reveal_type(x)
```

But I was hasty and failed to notice that the cause is that the first `if` doesn't construct a union type. You are correct that if I explicitly build a union of `Any| dict[str, str]`, the `isinstance` does narrow away the `Any`.

> What would you expect the revealed type to be in this case? If pyright favored strictness over practicality, the answer would be dict[str, str] | dict[Unknown, Unknown] because the type of x could be derived from Foo and dict.
Oof, of course.

On reflection, I think the actually-surprising behavior here to me is that pyright constructs union types with `Any` or `Unknown` instead of collapsing them, which has not been the case for gradual type systems I've worked with before; but I can understand that as a design decision.
